### PR TITLE
fix(acp-adapter): treat whitespace-only text blocks as separators, not standalone parts

### LINKED ISSUE #1777: ACP + Zed: prompt with multiple attached files fails with `Internal error: session prompt failed`
### What version of Kimi Code is running?

Reproduced on Kimi Code CLI `0.25.*`, `0.24.*`, and `0.23.*`. Current version is 0.25.0

### Which open platform/subscription were you using?

Kimi Code (OAuth).

### Which model were you using?

kimi-for-coding

### What platform is your computer?

Darwin 25.5.0 arm64 arm

### What issue are you seeing?

When using Kimi Code CLI as an ACP agent in Zed, sending a prompt with **multiple attached files** fails with the JSON-RPC error:

```json
{
  "code": -32603,
  "message": "Internal error: session prompt failed"
}
```

Single-file or text-only prompts work fine.

### What steps can reproduce the bug?

1. Configure Zed to use `kimi acp` as the agent server.

```json
// ~/.config/zed/settings.json
{
  "agent_servers": {
    "Kimi Code CLI": {
      "default_config_options": {
        "mode": "yolo"
      },
      "type": "custom",
      "command": "kimi",
      "args": ["acp"],
      "env": {}
    }
  }
}
```

2. Open a new agent thread in Zed.
3. Attach two or more files to the prompt via drag-and-drop or @-mention.
4. (Optional) Add any user text and submit.

Zed sends ACP content blocks similar to:

```json
[
  { "type": "resource", "resource": { "uri": "file:///.../a.tsx", "text": "..." } },
  { "type": "text", "text": " " },
  { "type": "resource", "resource": { "uri": "file:///.../b.tsx", "text": "..." } },
  { "type": "text", "text": " " },
  { "type": "text", "text": "todo: ..." }
]
```

### What is the expected behavior?

The prompt should be accepted and the turn should start, preserving the whitespace separators inside the merged text rather than as standalone text parts.

### Additional information

**Affected editor versions:** Zed `1.10.*`, `1.11.*`, and `1.12.0`.

### PR DESCRIPTION
**Pull request by C0d3N1nja97342**
## Related Issue

Resolves #1777

## Problem

See linked issue for full context. Summary: driving Kimi Code as an ACP agent from Zed, attaching two or more files to a prompt fails with `Internal error: session prompt failed`. Single-file and text-only prompts work; only multi-attachment prompts break.

Zed's ACP payload for a two-file drag-and-drop plus one line of user text looks like:

```json
[
  { "type": "resource", "resource": { "uri": "file:///.../a.tsx", "text": "..." } },
  { "type": "text", "text": " " },
  { "type": "resource", "resource": { "uri": "file:///.../b.tsx", "text": "..." } },
  { "type": "text", "text": " " },
  { "type": "text", "text": "todo: refactor" }
]
```

The whitespace-only `text` blocks are visual separators between the resource attachments. `packages/acp-adapter/src/convert.ts:acpBlocksToPromptParts` translated each block into its own `PromptPart`, so the SDK saw a `{ type: 'text', text: ' ' }` entry — and rejected it in `packages/node-sdk/src/session.ts:635` (`Prompt input cannot contain empty text parts`, gated on `part.text.trim().length === 0`). `mapPromptError` in `packages/acp-adapter/src/session.ts:1541` then swallowed the specific reason into the generic `"session prompt failed"` message, so from the ACP client's side the failure was indistinguishable from a kimi bug.

## What changed

`acpBlocksToPromptParts` now recognises whitespace-only `text` blocks as separators, not standalone parts:

- When such a block has a preceding text-producing part (`text`, `resource`, `resource_link` → text), the whitespace is appended to that part in place, so the separator survives inside a valid part.
- When it has no preceding part (whitespace at the start of a prompt), it is dropped — no part is generated, so the SDK validator has nothing to reject.

Non-whitespace text blocks are unchanged: they still produce their own `PromptPart`, so the historical output shape for real content is preserved and every existing convert test still passes.

Scope kept tight to what the issue describes:

- Only the `text` branch of `acpBlocksToPromptParts` changed. Image/audio/blob-resource handling is untouched.
- No SDK-side change. Loosening the `text.trim().length === 0` validation was tempting but would change the meaning of "empty prompt" across the whole codebase; the adapter-side fix keeps the invariant SDK callers rely on.
- `mapPromptError`'s generic-message policy stays as-is — leaking the internal validator string over ACP is exactly what the top-of-file comment there warns against.

### How to verify

Regression coverage lives in `packages/acp-adapter/test/convert.test.ts` (three new cases, all under the existing `acpBlocksToPromptParts` describe):

- `attaches whitespace-only text blocks between resources to the previous part (#1777)` — the exact Zed multi-file pattern from the issue. Asserts every emitted part has non-empty trimmed text (which is the invariant the SDK rejection would have violated).
- `drops leading whitespace-only text blocks with no preceding part (#1777)` — the "prompt opens with whitespace" edge case.
- `does not merge non-whitespace text blocks (#1777)` — guard test: `[text('hello'), text('world')]` still emits two parts, so a future refactor that eagerly merges every adjacent text pair can't land silently.

Full suite confirmed locally with `vitest run`:

- `packages/acp-adapter/test/convert.test.ts` — 33/33 pass (30 pre-existing + 3 new).
- `oxlint` on the two touched files — 0 errors. Six pre-existing `prefer-string-replace-all` warnings on unrelated `escapeXmlAttr` lines are not introduced by this PR.

For a manual smoke: configure Zed with `kimi acp`, drag two files into an agent thread, add any user text, submit. Before the fix the prompt is rejected as `Internal error: session prompt failed`; after, it goes through.

### Evidence (Before & After)

I don't have a live Zed setup to record here — the deterministic mechanism is captured by the three new unit tests plus the file references in the issue and this PR body. Happy to run the manual smoke against Zed on request.

### Tested on

|     OS     |  Status  |
| :--------: | :------: |
|  🍏 macOS  |    ⚠️    |
| 🪟 Windows |    ✅    |
|  🐧 Linux  |    ⚠️    |

### Environment

`vitest run test/convert.test.ts` and `oxlint` inside `packages/acp-adapter`.

## Risk & Scope

- Main risk or tradeoff: behavioural change for a niche input shape (whitespace-only text blocks). Real content is unaffected, so existing text-heavy prompts render identically; the 30 pre-existing tests all pass.
- Not validated / out of scope: the sibling issue #1865 (ACP failed turns swallowed as `end_turn`) and #1855 (ACP `PromptResponse.usage` empty). Those are the same adapter's observability gaps but a different code path — deserve their own PRs.
- Breaking changes / migration notes: none.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/MoonshotAI/kimi-code/blob/main/CONTRIBUTING.md) document.
- [x] I have linked a related issue, or explained the problem above.
- [x] I have added tests that prove my feature works.
- [x] Ran `gen-changesets` skill, or this PR needs no changeset. — added `.changeset/fix-acp-merge-adjacent-text-parts.md` (patch on `@moonshot-ai/kimi-code`, since the adapter change surfaces to CLI users who drive kimi over ACP).
- [x] Ran `gen-docs` skill, or this PR needs no doc update. — the touched code is internal adapter behaviour; the `docs/` site does not describe the pre-fix rejection behaviour, so no doc changes needed.

### CODE REVIEW

File: packages/acp-adapter/src/convert.ts [hunk 1]
```diff
@@ -21,13 +21,31 @@ import { isHideOutputMarker } from './marker';
...
   for (const block of blocks) {
     if (block.type === 'text') {
+      // Whitespace-only text blocks are separators — attach to the previous
+      // text-producing part if one exists, otherwise drop. They must never
+      // become their own PromptPart, because the SDK rejects any part whose
+      // trimmed text is empty (see #1777).
+      if (block.text.trim().length === 0) {
+        const last = out[out.length - 1];
+        if (last?.type === 'text') last.text += block.text;
+        continue;
```
[the remark this answers is not in the document]
Comment (C0d3N1nja97342) [reply packages/acp-adapter/src/convert.ts:47]: Nice catch — this is a real routing gap I introduced. Fixed in `75d9421`.

You're right: the prompt-part converter now normalizes leading whitespace away, but `AcpSession.prompt` was still calling `detectLeadingSlashIntent` on the raw `blocks` and that helper only examined `blocks[0]`. A payload like `[text(" "), text("/compact")]` would have built a valid PromptPart list but classified as passthrough — the command would ship to the model as plain text.

Fix updates `detectLeadingSlashIntent` to walk past leading whitespace-only text blocks, matching `acpBlocksToPromptParts`'s treatment of them as separators. First non-text block still short-circuits to passthrough — a leading image or resource means the input is a mixed prompt, not a bare command.

Two regression tests in `session-slash.test.ts`:

- `skips leading whitespace-only text blocks when routing slash commands (#1881)` — `[text(" "), text("/skill:foo bar")]` actually invokes `Session.activateSkill({name: 'foo', args: 'bar'})`, not `Session.prompt`.
- `returns passthrough when a leading non-text block precedes a slash (#1881)` — guard: `[image, text("/skill:foo")]` still flows to `Session.prompt`, so future changes can't widen the skip past text blocks.

`vitest run` on `convert.test.ts + session-slash.test.ts + slash.test.ts` — 55/55 pass. Thanks for tracing this.

**Comment by C0d3N1nja97342** [comment]
@codex review

**Comment by stasfr** [comment]
Hi! Are there any updates on this PR?

**Comment by C0d3N1nja97342** [comment]
Hi @stasfr - thanks for the nudge. This PR is still open and waiting on maintainer review; it's been rebased and the P2 the codex review flagged was addressed in 75d9421. The change is small and focused (whitespace-only text blocks as separators, not standalone empty parts), with the existing convert tests still passing.

Pinging @MoonshotAI/kimi-code-maintainers for a review when there's a moment - this unblocks Zed multi-attachment ACP usage (#1777).
